src/test/java/com/microsoft/sqlserver/jdbc/fedauth/PooledConnectionTest.java

```diff
@@ -200,17 +210,23 @@ private void testPooledConnectionMultiThread(long testingTimeInSeconds,
             Random rand = new Random();
             int numberOfThreadsForEachType = rand.nextInt(15) + 1; // 1 to 15
             for (int i = 0; i < numberOfThreadsForEachType; i++) {
-                new Thread(r1).start();
-                new Thread(r2).start();
-                new Thread(r3).start();
-            }
-
-            // sleep in order to catch exception from other threads if tests fail.
-            try {
-                Thread.sleep(TimeUnit.SECONDS.toMillis(60));
-
-            } catch (InterruptedException e) {
-                fail(e.getMessage());
+                Thread t1 = new Thread(r1);
+                Thread t2 = new Thread(r2);
+                Thread t3 = new Thread(r3);
+                t1.setUncaughtExceptionHandler(handler);
+                t2.setUncaughtExceptionHandler(handler);
+                t3.setUncaughtExceptionHandler(handler);
+                t1.start();
+                t2.start();
+                t3.start();
+                t1.join();
+                t2.join();
+                t3.join();

```

Comment:
We've already changed r1/r2/r3 to runnables, this should be done with an ExecutorService instead. The runnables above should also not be duplicated, they don't have different contents. Suggested Change:
```java
Runnable r = () -> {
    try (Connection connection2 = pc.getConnection()) {
        testUserName(connection2, azureUserName, authentication);
    } catch (SQLException e) {
        assertTrue(INVALID_EXCEPION_MSG + ": " + e.getMessage(),
                e.getMessage().contains(ERR_MSG_CONNECTION_CLOSED)
                        || e.getMessage().contains(ERR_MSG_CONNECTION_IS_CLOSED)
                        || e.getMessage().contains(ERR_MSG_HAS_CLOSED)
                        || e.getMessage().contains(ERR_MSG_HAS_BEEN_CLOSED)
                        || e.getMessage().contains(ERR_MSG_SOCKET_CLOSED));
    } catch (Exception e) {
        // do whatever error handling is meant to happen
    }
};

ExecutorService es = Executors.newFixedThreadPool(4);
es.execute(r);
es.execute(r);
es.execute(r);
es.shutdown();

```